### PR TITLE
[Evals API][3/n] scoring_functions / scoring meta-reference implementations

### PR DESCRIPTION
**Pull request by yanxi0830**
---continuation of https://github.com/meta-llama/llama-stack/pull/288

# TL;DR
- `scoring` depends on `datasetio` & `datasets`
- Enforce dataset_schema, client enforces dataset schema to follow specified column names/types
- Add clients for datasets / datasetio / scoring
- Basic meta-reference impl for scoring/score, scoring/score_batch

Next PR
- full evals benchmark task with generation
- LLM as judge scorer impl

# Tests
#### unit tests
```
PROVIDER_ID=test-meta PROVIDER_CONFIG=llama_stack/providers/tests/scoring/provider_config_example.yaml pytest -s llama_stack/providers/tests/scoring/test_scoring.py --tb=short --disable-warnings
```
image

```
PROVIDER_ID=test-meta PROVIDER_CONFIG=llama_stack/providers/tests/datasetio/provider_config_example.yaml pytest -s llama_stack/providers/tests/datasetio/test_datasetio.py --tb=short --disable-warnings
```

#### client tests
```
python -m llama_stack.apis.datasetio.client $LOCALHOST 5000
```
```
python -m llama_stack.apis.datasets.client $LOCALHOST 5000
```
```
python -m llama_stack.apis.scoring.client $LOCALHOST 5000
```
image

```
results={
  'equality': ScoringResult(score_rows=[
    {
      'score': 0.0
    },
    {
      'score': 1.0
    },
    {
      'score': 1.0
    },
    {
      'score': 0.0
    }
  ],
  aggregated_results={
    'accuracy': 0.5,
    'num_correct': 2.0,
    'num_total': 4
  })
}
```

### CODE REVIEW

File: llama_stack/apis/scoring_functions/scoring_functions.py [hunk 1]
```diff
@@ -34,8 +34,8 @@ class Parameter(BaseModel):
 
 
 @json_schema_type
-class CommonDef(BaseModel):
-    name: str
+class CommonFunctionDef(BaseModel):
+    identifier: str
```
Comment (yanxi0830) [inline llama_stack/apis/scoring_functions/scoring_functions.py:38]: all `Def` needs an `identifier`.

File: llama_stack/apis/scoring/scoring.py [hunk 2]
```diff
@@ -37,7 +38,10 @@ class Scoring(Protocol):
 
     @webmethod(route="/scoring/score_batch")
     async def score_batch(
-        self, dataset_id: str, scoring_functions: List[str]
+        self,
+        dataset_id: str,
+        scoring_functions: List[str],
+        save_results_dataset: bool = False,
```
Comment (yanxi0830) [inline llama_stack/apis/scoring/scoring.py:44]: flag to indicate whether we should/should not register the dataset

File: llama_stack/apis/scoring_functions/scoring_functions.py [hunk 3]
```diff
@@ -46,10 +46,11 @@ class CommonDef(BaseModel):
 
 
 @json_schema_type
-class DeterministicFunctionDef(CommonDef):
+class DeterministicFunctionDef(CommonFunctionDef):
```
Comment (ashwinb) [inline llama_stack/apis/scoring_functions/scoring_functions.py:49]: cc @raghotham about this name. this is the best we have for now to distinguish it from LLMJudgeFunctionDef. The obvious term "ScoringFunctionDef" is taken up by the Union type below so this needs to be a sub-type of that, so we need some kind of prefix. "Regular" is maybe another possibility, but it is also odd.

File: llama_stack/providers/impls/meta_reference/datasetio/datasetio.py [hunk 4]
```diff
@@ -123,7 +136,7 @@ async def get_rows_paginated(
         dataset_info = self.dataset_infos.get(dataset_id)
         dataset_info.dataset_impl.load()
 
-        if page_token is None:
+        if page_token is None or not page_token.isnumeric():
```
Comment (ashwinb) [inline llama_stack/providers/impls/meta_reference/datasetio/datasetio.py:139]: can that be an error if it is not numeric?

File: llama_stack/providers/impls/meta_reference/datasetio/datasetio.py [hunk 5]
```diff
@@ -88,6 +100,7 @@ def load(self) -> None:
             raise ValueError(f"Unsupported file type: {self.dataset_def.url}")
 
         self.df = df
+        self._validate_dataset_schema()
```
Comment (ashwinb) [inline llama_stack/providers/impls/meta_reference/datasetio/datasetio.py:103]: nitpick -- you should not set `self.df` before it is validated. so instead do: `self.df = self._validate_dataset_schema(df)`

File: llama_stack/providers/impls/meta_reference/scoring/scoring.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        dataset_def = await self.datasets_api.get_dataset(dataset_identifier=dataset_id)
+        if not dataset_def.dataset_schema or len(dataset_def.dataset_schema) == 0:
+            raise ValueError(
+                f"Dataset {dataset_id} does not have a schema defined. Please define a schema for the dataset."
+            )
+
+        for required_column in ["generated_answer", "expected_answer", "input_query"]:
+            if required_column not in dataset_def.dataset_schema:
+                raise ValueError(
+                    f"Dataset {dataset_id} does not have a '{required_column}' column. Please make sure '{required_column}' column is in the dataset."
```
Comment (ashwinb) [inline llama_stack/providers/impls/meta_reference/scoring/scoring.py:62]: Second sentence can be dropped in all these error messages since it is repetitive.

File: llama_stack/providers/impls/meta_reference/scoring/scoring.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                )
+
+    async def score_batch(
+        self,
+        dataset_id: str,
+        scoring_functions: List[str],
+        save_results_dataset: bool = False,
+    ) -> ScoreBatchResponse:
+        await self.validate_scoring_input_dataset_schema(dataset_id=dataset_id)
+        rows_paginated = await self.datasetio_api.get_rows_paginated(
```
Comment (ashwinb) [inline llama_stack/providers/impls/meta_reference/scoring/scoring.py:76]: nit: name it as `all_rows`?

File: llama_stack/providers/impls/meta_reference/scoring/scoring.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            raise NotImplementedError("Save results dataset not implemented yet")
+
+        return ScoreBatchResponse(
+            results=res.results,
+        )
+
+    async def score(
+        self, input_rows: List[Dict[str, Any]], scoring_functions: List[str]
+    ) -> ScoreResponse:
+        res = {}
```
Comment (ashwinb) [inline llama_stack/providers/impls/meta_reference/scoring/scoring.py:95]: assert that all functions are registered and known?

**Review by ashwinb — APPROVED** [review]
I think this is totally land-able. @raghotham thoughts?

File: llama_stack/distribution/routers/routers.py [hunk 9]
```diff
@@ -192,3 +193,56 @@ async def get_rows_paginated(
...
+        scoring_functions: List[str],
+        save_results_dataset: bool = False,
+    ) -> ScoreBatchResponse:
+        res = {}
+        for fn_identifier in scoring_functions:
+            score_response = await self.routing_table.get_provider_impl(
+                fn_identifier
+            ).score_batch(
+                dataset_id=dataset_id,
+                scoring_functions=[fn_identifier],
```
[the remark this answers is not in the document]
Comment (yanxi0830) [reply llama_stack/distribution/routers/routers.py:223]: The reason for the for loop is b/c we could have a list of scoring functions e.g. [meta-reference-accuracy, braintrust-factuality, ...], and we need the router to .call `get_provider_impl` to route to the specific provider (meta-reference).

File: llama_stack/apis/scoring_functions/scoring_functions.py (hunk 3 above)
Comment (yanxi0830) [reply llama_stack/apis/scoring_functions/scoring_functions.py:49]: Another thing is that we could technically define a braintrust's provider LLM As Judge scoring function using a `DeterministicFunctionDef`, so there's some ambiguity when choosing b/w "deterministic" v.s. "judge" types.

File: llama_stack/providers/impls/meta_reference/scoring/scoring.py (hunk 6 above)
Comment (yanxi0830) [reply llama_stack/providers/impls/meta_reference/scoring/scoring.py:62]: Haha, got an elaborate auto-complete

File: llama_stack/apis/scoring_functions/scoring_functions.py (hunk 3 above)
Comment (ashwinb) [reply llama_stack/apis/scoring_functions/scoring_functions.py:49]: @yanxi0830 yes totally. we need to kill Deterministic; let me think for a bit

File: llama_stack/apis/scoring/scoring.py [hunk 10]
```diff
@@ -18,13 +18,14 @@
...
-    dataset_id: str
+    dataset_id: Optional[str] = None
+    results: Dict[str, ScoringResult]
 
 
 @json_schema_type
 class ScoreResponse(BaseModel):
     # each key in the dict is a scoring function name
-    results: List[Dict[str, ScoringResult]]
+    results: Dict[str, ScoringResult]
```
Comment (ashwinb) [inline llama_stack/apis/scoring/scoring.py:28]: why this change? wouldn't you have one dict per one input row?

Comment (yanxi0830) [reply llama_stack/apis/scoring/scoring.py:28]: I had made it s.t. we perform score aggregation (e.g. avg, sum) etc as part of scorer. We can add it back to deligate the aggregation part to the client.

Comment (yanxi0830) [reply llama_stack/apis/scoring/scoring.py:28]: Ok, I think aggregation on client might be better, gives a chance for analyzing each example's scores.

Comment (yanxi0830) [reply llama_stack/apis/scoring/scoring.py:36]: Updated scoring response output

```
results={
  'equality': ScoringResult(score_rows=[
    {
      'score': 0.0
    },
    {
      'score': 1.0
    },
    {
      'score': 1.0
    },
    {
      'score': 0.0
    }
  ],
  aggregated_results={
    'accuracy': 0.5,
    'num_correct': 2.0,
    'num_total': 4
  })
}
```

File: llama_stack/apis/scoring_functions/scoring_functions.py (hunk 3 above)
Comment (yanxi0830) [reply llama_stack/apis/scoring_functions/scoring_functions.py:49]: `ScoringFunctionDef` / `ScoringFunctionWithInferenceDef`?

- ScoringFunctionDef for general scoring functions & including braintrust LLMJudge
- ScoringFunctionWithInferenceDef for LLM judge scoring where the judge is a model needing LlamaStack inference

Comment (yanxi0830) [reply llama_stack/apis/scoring_functions/scoring_functions.py]: Synced offline, updated to single `ScoringFunctionDef ` w/ optional `context` param.
